src/NewChain/ChainVerifier.ts

```diff
@@ -0,0 +1,414 @@
+  private originWeb3: Web3;
+
+  private auxiliaryWeb3: Web3;
+
+  private originChainIdentifier: string;
+
+  private auxiliaryChainId: string;
+
+  private mosaicConfig: MosaicConfig;
+
+  private mosaicContract: MosaicContracts;
+
+  private contractAddresses: ContractAddresses;
+
+  private abiBinProvider: AbiBinProvider;
+
+  /**
+   * ChainVerifier constructor.
+   *
+   * @param {string} originWebSocket Origin chain web3 endpoint.
+   * @param {string} auxiliaryWebSocket Auxiliary chain web3 endpoint.
+   * @param {string} originChainIdentifier Origin chain identifier.
+   * @param {string} auxiliaryChainId Auxiliary chain id.
+   */
+  public constructor(
+    originWebSocket: string,
+    auxiliaryWebSocket: string,
+    originChainIdentifier: string,
+    auxiliaryChainId: string,
+  ) {
+    this.originWeb3 = new Web3(originWebSocket);
+    this.auxiliaryWeb3 = new Web3(auxiliaryWebSocket);
+    this.originChainIdentifier = originChainIdentifier;
+    this.auxiliaryChainId = auxiliaryChainId;
+    this.mosaicConfig = MosaicConfig.from(this.originChainIdentifier);
+    this.mosaicContract = new MosaicContracts(this.originWeb3, this.auxiliaryWeb3);
+    this.contractAddresses = this.mosaicConfig.auxiliaryChains[this.auxiliaryChainId]
+      .contractAddresses;
+    this.abiBinProvider = new AbiBinProvider();
+  }
+
+  /**
+   * Public method verify which encapsulates chain verification logic.
+   *
+   * @returns {Promise<void>}
+   */
+  public async verify(): Promise<void> {
+    Logger.info('Starting chain verification!!!');
+
+    await this.verifyContractsBin();
+    await this.verifyGateway();
+    await this.verifyCoGateway();
+    await this.verifyOriginAnchor();
+    await this.verifyAuxiliaryAnchor();
+    await this.verifyOSTPrime();
+
+    Logger.info('Successfully completed chain verification!!!');
+  }
+
+  /**
+   * This method does verification of deployed contracts BIN.
+   * Note: Runtime bytecode is what's actually stored at the contract address.
+   * It doesn't include the constructor function or initialization code (which the bytecode does).
+   * 0x is stripped and checked if runtime bytecode is substring of sent bytecode
+   *
+   * @returns {Promise<void>}
+   */
+  private async verifyContractsBin(): Promise<void> {
+    const deployedGatewayBin = await this.originWeb3.eth.getCode(
+      this.contractAddresses.origin.ostEIP20GatewayAddress,
+    );
+
+    const messageBusLinkInfo = {
+      address: this.mosaicConfig.originChain.contractAddresses.messageBusAddress,
+      name: 'MessageBus',
+    };
+    const gatewayLibLinkInfo = {
+      address: this.mosaicConfig.originChain.contractAddresses.gatewayLibAddress,
+      name: 'GatewayLib',
+    };
+    const gatewayLinkedBin = this.abiBinProvider.getLinkedBIN(
+      'EIP20Gateway',
+      messageBusLinkInfo,
+      gatewayLibLinkInfo,
+    );
+    if (gatewayLinkedBin.toLowerCase().indexOf(deployedGatewayBin.slice(2).toLowerCase()) === -1) {
```

Comment:
Runtime bytecode is what's actually stored at the contract address. It doesn't include the constructor function or initialization code (which the sent bytecode does).
Because of above code hash will not match exactly.